src/node/txreconciliation.cpp

```diff
@@ -121,26 +201,181 @@ class TxReconciliationTracker::Impl
+
+                size_t local_set_size = recon_state.m_local_set.size();
+
+                LogPrint(BCLog::NET, "Initiate reconciliation with peer=%d with the following params: " /* Continued */
+                                     "local_set_size=%i\n",
+                         peer_id, local_set_size);
+
+                // In future, Q could be recomputed after every reconciliation based on the
+                // set differences. For now, it provides good enough results without recompute
+                // complexity, but we communicate it here to allow backward compatibility if
+                // the value is changed or made dynamic.
+                return std::make_pair(local_set_size, Q * Q_PRECISION);
+            }
+        }
+        return std::nullopt;
+    }
+
+    size_t GetPeerSetSize(NodeId peer_id) const EXCLUSIVE_LOCKS_REQUIRED(!m_txreconciliation_mutex)
+    {
+        AssertLockNotHeld(m_txreconciliation_mutex);
+        LOCK(m_txreconciliation_mutex);
+        assert(IsPeerRegistered(peer_id));
+        const auto& recon_state = std::get<TxReconciliationState>(m_states.find(peer_id)->second);
+
+        return recon_state.m_local_set.size();
+    }
+
     void ForgetPeer(NodeId peer_id) EXCLUSIVE_LOCKS_REQUIRED(!m_txreconciliation_mutex)
     {
         AssertLockNotHeld(m_txreconciliation_mutex);
         LOCK(m_txreconciliation_mutex);
         if (m_states.erase(peer_id)) {
+            m_queue.erase(std::remove(m_queue.begin(), m_queue.end(), peer_id), m_queue.end());
             LogPrintLevel(BCLog::TXRECONCILIATION, BCLog::Level::Debug, "Forget txreconciliation state of peer=%d\n", peer_id);
         }
     }
 
-    bool IsPeerRegistered(NodeId peer_id) const EXCLUSIVE_LOCKS_REQUIRED(!m_txreconciliation_mutex)
+    bool IsPeerRegistered(NodeId peer_id) const EXCLUSIVE_LOCKS_REQUIRED(m_txreconciliation_mutex)
     {
-        AssertLockNotHeld(m_txreconciliation_mutex);
-        LOCK(m_txreconciliation_mutex);
+        AssertLockHeld(m_txreconciliation_mutex);
         auto recon_state = m_states.find(peer_id);
         return (recon_state != m_states.end() &&
                 std::holds_alternative<TxReconciliationState>(recon_state->second));
     }
+
+    bool IsPeerRegisteredExternal(NodeId peer_id) const EXCLUSIVE_LOCKS_REQUIRED(!m_txreconciliation_mutex)
+    {
+        AssertLockNotHeld(m_txreconciliation_mutex);
+        LOCK(m_txreconciliation_mutex);
+        return IsPeerRegistered(peer_id);
+    }
+
+    bool ShouldFloodTo(NodeId peer_id, const uint256& wtxid) const EXCLUSIVE_LOCKS_REQUIRED(!m_txreconciliation_mutex)
+    {
+        AssertLockNotHeld(m_txreconciliation_mutex);
+        LOCK(m_txreconciliation_mutex);
+        if (!IsPeerRegistered(peer_id)) return true;
+        const auto& recon_state = std::get<TxReconciliationState>(m_states.find(peer_id)->second);
+
+        // We assume that reconciliation is always initiated from inbound to outbound to avoid
```

Comment:
If you can use the same wording than the BIP here "The initiator of the P2P connection assumes the role of reconciliation initiator" otherwise it's obscure how the initiator/responder overlaps with connection direction.

Reply:
This comment was outdated, i think it's not needed here at all... deleting for now.